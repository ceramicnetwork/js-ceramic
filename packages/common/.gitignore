node_modules/
coverage/
lib/
src/generated/
