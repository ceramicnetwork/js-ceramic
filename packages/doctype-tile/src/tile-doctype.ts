import jsonpatch from 'fast-json-patch'

import { encode as base64Encode } from '@ethersproject/base64'
import { randomBytes } from '@ethersproject/random'

import { DID } from 'dids'
import {
    Doctype,
    DoctypeConstructor,
    DoctypeStatic,
    CeramicCommit,
    CommitHeader,
    DocOpts,
    DocParams,
    Context,
    GenesisHeader,
    GenesisCommit,
    UnsignedCommit,
} from "@ceramicnetwork/common"

export const DOCTYPE_NAME = 'tile'

/**
 * Tile doctype parameters
 */
export interface TileParams extends DocParams {
    content?: Record<string, unknown>;
}

/**
 * Tile doctype implementation
 */
@DoctypeStatic<DoctypeConstructor<TileDoctype>>()
export class TileDoctype extends Doctype {

    /**
     * Change existing Tile doctype
     * @param params - Change parameters
     * @param opts - Initialization options
     */
    async change(params: TileParams, opts: DocOpts = {}): Promise<void> {
        if (this.context.did == null) {
            throw new Error('No DID authenticated')
        }

        const updateCommit = await TileDoctype._makeCommit(this, this.context.did, params.content, params.metadata?.controllers, params.metadata?.schema)
        const updated = await this.context.api.applyCommit(this.id.toString(), updateCommit, opts)
        this.state = updated.state
    }

    /**
     * Create Tile doctype
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async create(params: TileParams, context: Context, opts?: DocOpts): Promise<TileDoctype> {
        const { content, metadata } = params
        const commit = await TileDoctype.makeGenesis({ content, metadata }, context)
        return context.api.createDocumentFromGenesis<TileDoctype>(DOCTYPE_NAME, commit, opts)
    }

    /**
     * Creates genesis commit
     * @param params - Create parameters
     * @param context - Ceramic context
     */
    static async makeGenesis<T extends CeramicCommit>(params: DocParams, context: Context): Promise<T> {
        // If 'deterministic' is undefined, default to creating document uniquely
        const unique = params.deterministic ? '0' : base64Encode(randomBytes(12))

        const metadata: GenesisHeader = params.metadata || { controllers: [] }
        if (!metadata.controllers || metadata.controllers.length === 0) {
            if (params.content && context.did) {
                metadata.controllers = [context.did.id]
            } else {
                throw new Error('No controllers specified')
            }
        }

        if (metadata.controllers.length !== 1) {
            throw new Error('Exactly one controller must be specified')
        }

        const commit: GenesisCommit = { data: params.content, header: metadata, unique }
        return (params.content ? await TileDoctype._signDagJWS(commit, context.did, metadata.controllers[0]): commit) as T
    }

    /**
     * Make change commit
     * @param doctype - Tile doctype instance
     * @param did - DID instance
     * @param newContent - New context
     * @param newControllers - New controllers
     * @param schema - New schema ID
     * @private
     */
    static async _makeCommit(doctype: Doctype, did: DID, newContent: any, newControllers?: string[], schema?: string): Promise<CeramicCommit> {
        const header: CommitHeader = {
            ...schema != null && { schema: schema },
            ...newControllers != null && { controllers: newControllers },
        }

        if (newContent == null) {
            newContent = doctype.content
        }

        if (header.controllers && header.controllers.length !== 1) {
            throw new Error('Exactly one controller must be specified')
        }

        const patch = jsonpatch.compare(doctype.content, newContent)
        const commit: UnsignedCommit = { header, data: patch, prev: doctype.tip, id: doctype.state.log[0].cid }
        return TileDoctype._signDagJWS(commit, did, doctype.controllers[0])
    }

    /**
     * Sign Tile commit
     * @param did - DID instance
     * @param commit - Commit to be signed
     * @param controller - Controller
     * @private
     */
    static async _signDagJWS(commit: CeramicCommit, did: DID, controller: string): Promise<CeramicCommit> {
        // check for DID and authentication
        if (did == null || !did.authenticated) {
            throw new Error('No DID authenticated')
        }
        return did.createDagJWS(commit, { did: controller })
    }

}
