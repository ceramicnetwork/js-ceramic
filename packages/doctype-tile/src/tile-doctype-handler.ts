import CID from 'cids'

import * as didJwt from 'did-jwt'
import base64url from 'base64url'
import { PublicKey } from 'did-resolver'

import jsonpatch from 'fast-json-patch'
import cloneDeep from 'lodash.clonedeep'

import { TileDoctype, TileParams } from "./tile-doctype"
import {
    AnchorProof,
    AnchorRecord,
    AnchorStatus,
    Context,
    DocOpts,
    DocState,
    RecordType,
    DoctypeConstructor,
    DoctypeHandler,
    DoctypeUtils,
    SignatureStatus
} from "@ceramicnetwork/common"

const DOCTYPE = 'tile'

/**
 * Tile doctype handler implementation
 */
export class TileDoctypeHandler implements DoctypeHandler<TileDoctype> {
    /**
     * Gets doctype name
     */
    get name(): string {
        return DOCTYPE
    }

    /**
     * Gets doctype class
     */
    get doctype(): DoctypeConstructor<TileDoctype> {
        return TileDoctype
    }

    /**
     * Create new Tile doctype instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    async create(params: TileParams, context: Context, opts?: DocOpts): Promise<TileDoctype> {
        return TileDoctype.create(params, context, opts)
    }

    /**
     * Applies record (genesis|signed|anchor)
     * @param record - Record
     * @param cid - Record CID
     * @param context - Ceramic context
     * @param state - Document state
     */
    async applyRecord(record: any, cid: CID, context: Context, state?: DocState): Promise<DocState> {
        if (state == null) {
            // apply genesis
            return this._applyGenesis(record, cid, context)
        }

        if (record.proof) {
            const proofRecord = (await context.ipfs.dag.get(record.proof)).value;
            return this._applyAnchor(record, proofRecord, cid, state);
        }

        return this._applySigned(record, cid, state, context)
    }

    /**
     * Applies genesis record
     * @param record - Genesis record
     * @param cid - Genesis record CID
     * @param context - Ceramic context
     * @private
     */
    async _applyGenesis(record: any, cid: CID, context: Context): Promise<DocState> {
        let payload = record
        const isSigned = DoctypeUtils.isSignedRecord(record)
        if (isSigned) {
            payload = (await context.ipfs.dag.get(record.link)).value
            await this._verifySignature(record, context, payload.header.controllers[0])
        } else if (payload.data !== null) {
            throw Error('Genesis record with contents should always be signed')
        }
        return {
            doctype: DOCTYPE,
            content: payload.data,
            metadata: payload.header,
            signature: isSigned? SignatureStatus.SIGNED : SignatureStatus.GENESIS,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            log: [{ cid, type: RecordType.GENESIS }]
        }
    }

    /**
     * Applies signed record
     * @param record - Signed record
     * @param cid - Signed record CID
     * @param state - Document state
     * @param context - Ceramic context
     * @private
     */
    async _applySigned(record: any, cid: CID, state: DocState, context: Context): Promise<DocState> {
        await this._verifySignature(record, context, state.metadata.controllers[0])

        const payload = (await context.ipfs.dag.get(record.link)).value
        if (!payload.id.equals(state.log[0].cid)) {
            throw new Error(`Invalid docId ${payload.id}, expected ${state.log[0].cid}`)
        }

        const nextState = cloneDeep(state)

        nextState.signature = SignatureStatus.SIGNED
        nextState.anchorStatus = AnchorStatus.NOT_REQUESTED

        const nonce = payload.header.nonce
        const squash = nonce > 0 && state.next
        if (squash) {
            nextState.log[nextState.log.length-1] = { cid, type: RecordType.SIGNED }
            nextState.next = {
                content: jsonpatch.applyPatch(state.next.content, payload.data).newDocument
            }
        } else {
            nextState.log.push({ cid, type: RecordType.SIGNED })
            nextState.next = {
                content: jsonpatch.applyPatch(state.content, payload.data).newDocument
            }
        }

        if (payload.header) {
            nextState.next.metadata = {...nextState.metadata, ...payload.header}
        }

        if (nonce) {
            nextState.next.metadata = { ...nextState.next.metadata, nonce: nonce }
        }
        return nextState
    }

    /**
     * Applies anchor record
     * @param record - Anchor record
     * @param proof - Anchor record proof
     * @param cid - Anchor record CID
     * @param state - Document state
     * @private
     */
    async _applyAnchor(record: AnchorRecord, proof: AnchorProof, cid: CID, state: DocState): Promise<DocState> {
        state.log.push({ cid, type: RecordType.ANCHOR })
        let content = state.content
        let metadata = state.metadata

        if (state.next?.content) {
            content = state.next.content
            delete state.next.content
        }

        if (state.next?.metadata) {
            metadata = state.next.metadata
            delete state.next.metadata
        }

        delete state.next

        return {
            ...state, content, metadata, anchorStatus: AnchorStatus.ANCHORED, anchorProof: proof,
        }
    }

    /**
     * Verifies record signature
     * @param record - Record to be verified
     * @param context - Ceramic context
     * @param did - DID value
     * @private
     */
    async _verifySignature(record: any, context: Context, did: string): Promise<void> {
        const { payload, signatures } = record
        const { signature,  protected: _protected } = signatures[0]

        const decodedHeader = JSON.parse(base64url.decode(_protected))
        const { kid } = decodedHeader
        if (!kid.startsWith(did)) {
            throw new Error(`Signature was made with wrong DID. Expected: ${did}, got: ${kid}`)
        }

        const { publicKey } = await context.resolver.resolve(kid)
        const jws = [_protected, payload, signature].join('.')
        try {
            await this.verifyJWS(jws, publicKey)
        } catch (e) {
            throw new Error('Invalid signature for signed record. ' + e)
        }
    }

    /**
     * Verifies JWS token
     * @param jws - JWS token
     * @param pubkeys - public key(s)
     */
    async verifyJWS(jws: string, pubkeys: PublicKey[]): Promise<void> {
        await didJwt.verifyJWS(jws, pubkeys)
    }

}
