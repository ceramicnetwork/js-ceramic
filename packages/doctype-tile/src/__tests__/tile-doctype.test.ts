import CID from 'cids'

import dagCBOR from "ipld-dag-cbor"

import { DID } from 'dids'
import { Resolver } from "did-resolver"
import ThreeIdResolver from '@ceramicnetwork/3id-did-resolver'
import KeyDidResolver from '@ceramicnetwork/key-did-resolver'
import { TileDoctypeHandler } from '../tile-doctype-handler'

import { TileDoctype } from "../tile-doctype"
import { AnchorRecord, Context, DoctypeUtils, SignedRecordDTO } from "@ceramicnetwork/common"

jest.mock('did-jwt', () => ({
  // TODO - We should test for when this function throws as well
  verifyJWS: (): void => { return }
}))

const cloneDeep = require('lodash.clonedeep') // eslint-disable-line @typescript-eslint/no-var-requires
const { sha256 } = require('js-sha256') // eslint-disable-line @typescript-eslint/no-var-requires
const hash = (data: string): CID => new CID(1, 'sha2-256', Buffer.from('1220' + sha256(data), 'hex'))

const FAKE_CID_1 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = new CID('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_3 = new CID('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_4 = new CID('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlsu')

// did:3:bafyasdfasdf

const RECORDS = {
  genesis: { header: { controllers: [ 'did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki' ] }, data: { much: 'data' }, unique: '0' },
  genesisGenerated: {
    "jws": {
      "payload": "bbbb",
      "signatures": [
        {
          "protected": "eyJraWQiOiJkaWQ6MzprMnQ2d3lmc3U0cGcwdDJuNGo4bXMzczMzeHNncWpodHRvMDRtdnE4dzVhMnY1eG80OGlkeXozOGw3eWRraT92ZXJzaW9uPTAjc2lnbmluZyIsImFsZyI6IkVTMjU2SyJ9",
          "signature": "cccc"
        }
      ],
      "link": new CID("bafyreiau5pqllna6pewhp3w2hbvohxxeqsmffnwf6o2fwoln4ubbc6fldq")
    },
    "linkedBlock": {
      "data": {
        "much": "data"
      },
      "header": {
        "controllers": [
          "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki"
        ]
      },
      "unique": "0",
    }
  },
  r1: {
    desiredContent: { much: 'data', very: 'content' },
    record: {
      "jws": {
        "payload": "bbbb",
        "signatures": [
          {
            "protected": "eyJraWQiOiJkaWQ6MzprMnQ2d3lmc3U0cGcwdDJuNGo4bXMzczMzeHNncWpodHRvMDRtdnE4dzVhMnY1eG80OGlkeXozOGw3eWRraT92ZXJzaW9uPTAjc2lnbmluZyIsImFsZyI6IkVTMjU2SyJ9",
            "signature": "cccc"
          }
        ],
        "link": "bafyreia6chsgnfihmdrl2d36llrfevc6xgmgzryi3ittdg3j5ohdifb7he"
      },
      "linkedPayload": {
        "id": "bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu",
        "data": [
          {
            "op": "add",
            "path": "/very",
            "value": "content"
          }
        ],
        "prev": "bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu",
        "header": {},
      }
    }
  },
  r2: { record: { proof: FAKE_CID_4 } },
  proof: {
    blockNumber: 123456,
    chainId: 'fakechain:123',
  }
}

const serialize = (data: any): any => {
  if (Array.isArray(data)) {
    const serialized = []
    for (const item of data) {
      serialized.push(serialize(item))
    }
    return serialized
  }
  if (!CID.isCID(data) && typeof data === "object") {
    const serialized: Record<string, any> = {}
    for (const prop in data) {
      serialized[prop] = serialize(data[prop])
    }
    return serialized
  }
  if (CID.isCID(data)) {
    return data.toString()
  }
  return data
}

describe('TileDoctypeHandler', () => {
  let did: DID;
  let tileDoctypeHandler: TileDoctypeHandler;
  let context: Context;

  beforeAll(() => {
    did = new DID()
    did.createJWS = jest.fn(async () => {
      // fake jws
      return { payload: 'bbbb', signatures: [{ protected: 'eyJraWQiOiJkaWQ6MzprMnQ2d3lmc3U0cGcwdDJuNGo4bXMzczMzeHNncWpodHRvMDRtdnE4dzVhMnY1eG80OGlkeXozOGw3eWRraT92ZXJzaW9uPTAjc2lnbmluZyIsImFsZyI6IkVTMjU2SyJ9', signature: 'cccc'}]}
    })
    did._id = 'did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki'

    const recs: Record<string, any> = {}
    const ipfs = {
      dag: {
        put(rec: any, cid?: CID): any {
          if (cid) {
            recs[cid.toString()] = { value: rec }
            return cid
          }
          // stringify as a way of doing deep copy
          const clone = cloneDeep(rec)
          const c = hash(JSON.stringify(clone))
          recs[c.toString()] = { value: clone }
          return c
        },
        get(cid: any): any {
          return recs[cid.toString()]
        }
      }
    }

    const threeIdResolver = ThreeIdResolver.getResolver({
      loadDocument: (): any => {
        return Promise.resolve({
          content: {
            "publicKeys": {
              "signing": "zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
              "encryption": "z6LSfQabSbJzX8WAm1qdQcHCHTzVv8a2u6F7kmzdodfvUCo9"
            }
          }
        })
      },
      createDocument: (): any => { return null }
    })

    const api = { getSupportedChains: jest.fn(async () => {return ["fakechain:123"]}) }
    const keyDidResolver = KeyDidResolver.getResolver()
    context = {
      did,
      ipfs: ipfs,
      anchorService: null,
      resolver: new Resolver({
        ...threeIdResolver,
        ...keyDidResolver,
      }),
      api
    }
  })

  beforeEach(() => {
    tileDoctypeHandler = new TileDoctypeHandler()
  })

  it('is constructed correctly', async () => {
    expect(tileDoctypeHandler.name).toEqual('tile')
  })

  it('makes genesis record correctly', async () => {
    const record1 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] }, deterministic: true }, context)
    expect(record1).toBeDefined()

    const { jws, linkedBlock } = record1 as SignedRecordDTO
    expect(jws).toBeDefined()
    expect(linkedBlock).toBeDefined()

    const payload = dagCBOR.util.deserialize(linkedBlock)

    const serialized = { jws: serialize(jws), linkedBlock: serialize(payload)}

    const expected1 = await did.createDagJWS(RECORDS.genesis)
    expect(expected1).toBeDefined()

    const { jws: eJws, linkedBlock: eLinkedBlock } = expected1
    const ePayload = dagCBOR.util.deserialize(eLinkedBlock)
    const signed = { jws: serialize(eJws), linkedBlock: serialize(ePayload)}

    expect(serialized).toEqual(signed)
  })

  it('creates genesis records uniquely by default', async () => {
    const record1 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] } }, context)
    const record2 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] } }, context)

    expect(record1).not.toEqual(record2)
  })

  it('creates genesis records deterministically if deterministic:true is specified', async () => {
    const record1 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] }, deterministic: true }, context)
    const record2 = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] }, deterministic: true }, context )

    expect(record1).toEqual(record2)
  })

  it('creates genesis records without DID if content is undefined', async () => {
    expect(TileDoctype.makeGenesis({ content: 'asdf', metadata: { controllers: [did.id] } }, {})).rejects.toThrow('No DID authenticated')
    const record1 = await TileDoctype.makeGenesis({ metadata: { controllers: [did.id] } }, {})

    expect(record1).toBeDefined()
  })

  it('applies genesis record correctly', async () => {
    const tileHandler = new TileDoctypeHandler()

    const record = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(record, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(record.linkedBlock)
    await context.ipfs.dag.put(payload, record.jws.link)

    const docState = await tileHandler.applyRecord(record.jws, FAKE_CID_1, context)
    expect(docState).toMatchSnapshot()
  })

  it('makes signed record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    await context.ipfs.dag.put(RECORDS.genesisGenerated.jws, FAKE_CID_1)

    await context.ipfs.dag.put(RECORDS.genesisGenerated.linkedBlock, RECORDS.genesisGenerated.jws.link)

    const state = await tileDoctypeHandler.applyRecord(RECORDS.genesisGenerated.jws, FAKE_CID_1, context)
    const doctype = new TileDoctype(state, context)

    await expect(TileDoctype._makeRecord(doctype, null, RECORDS.r1.desiredContent)).rejects.toThrow(/No DID/)

    const record = await TileDoctype._makeRecord(doctype, did, RECORDS.r1.desiredContent) as SignedRecordDTO
    const { jws: rJws, linkedBlock: rLinkedBlock} = record
    const rPayload = dagCBOR.util.deserialize(rLinkedBlock)
    expect({ jws: serialize(rJws), linkedPayload: serialize(rPayload)}).toEqual(RECORDS.r1.record)
  })

  it('applies signed record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    const genesisRecord = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(genesisRecord, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(genesisRecord.linkedBlock)
    await context.ipfs.dag.put(payload, genesisRecord.jws.link)

    // apply genesis
    let state = await tileDoctypeHandler.applyRecord(genesisRecord.jws, FAKE_CID_1, context)

    const doctype = new TileDoctype(state, context)
    const signedRecord = await TileDoctype._makeRecord(doctype, did, RECORDS.r1.desiredContent) as SignedRecordDTO

    await context.ipfs.dag.put(signedRecord, FAKE_CID_2)

    const sPayload = dagCBOR.util.deserialize(signedRecord.linkedBlock)
    await context.ipfs.dag.put(sPayload, signedRecord.jws.link)

    // apply signed
    state = await tileDoctypeHandler.applyRecord(signedRecord.jws, FAKE_CID_2, context, state)
    expect(state).toMatchSnapshot()
  })

  it('squashes record correctly', async () => {
    const deepCopy = o => DoctypeUtils.deserializeState(DoctypeUtils.serializeState(o))
    const tileDoctypeHandler = new TileDoctypeHandler()

    const genesisRecord = await TileDoctype.makeGenesis({ content: { test: 'data' }, metadata: { controllers: [did.id] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(genesisRecord, FAKE_CID_1)
    const payload = dagCBOR.util.deserialize(genesisRecord.linkedBlock)
    await context.ipfs.dag.put(payload, genesisRecord.jws.link)
    // apply genesis
    const genesisState = await tileDoctypeHandler.applyRecord(genesisRecord.jws, FAKE_CID_1, context)

    // make a first update
    let doctype = new TileDoctype(genesisState, context)
    const signedRecord1 = await TileDoctype._makeRecord(doctype, did, { other: { obj: 'content' } }) as SignedRecordDTO

    await context.ipfs.dag.put(signedRecord1, FAKE_CID_2)
    const sPayload1 = dagCBOR.util.deserialize(signedRecord1.linkedBlock)
    await context.ipfs.dag.put(sPayload1, signedRecord1.jws.link)
    // apply signed
    const state1 = await tileDoctypeHandler.applyRecord(signedRecord1.jws, FAKE_CID_2, context, deepCopy(genesisState))

    // make a second update that squashes the first
    doctype = new TileDoctype(state1, context)
    const signedRecord2 = await TileDoctype._makeRecord(doctype, did, { other: { obj2: 'fefe' } }) as SignedRecordDTO

    await context.ipfs.dag.put(signedRecord2, FAKE_CID_3)
    const sPayload2 = dagCBOR.util.deserialize(signedRecord2.linkedBlock)
    await context.ipfs.dag.put(sPayload2, signedRecord2.jws.link)

    // apply signed
    const state2 = await tileDoctypeHandler.applyRecord(signedRecord2.jws, FAKE_CID_3, context, deepCopy(genesisState))

    expect(state2).toMatchSnapshot()
  })

  it('throws error if record signed by wrong DID', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    const genesisRecord = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: ['did:3:fake'] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(genesisRecord, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(genesisRecord.linkedBlock)
    await context.ipfs.dag.put(payload, genesisRecord.jws.link)

    await expect(tileDoctypeHandler.applyRecord(genesisRecord.jws, FAKE_CID_1, context)).rejects.toThrow(/wrong DID/)
  })

  it('applies anchor record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    const genesisRecord = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(genesisRecord, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(genesisRecord.linkedBlock)
    await context.ipfs.dag.put(payload, genesisRecord.jws.link)

    // apply genesis
    let state = await tileDoctypeHandler.applyRecord(genesisRecord.jws, FAKE_CID_1, context)

    const doctype = new TileDoctype(state, context)
    const signedRecord = await TileDoctype._makeRecord(doctype, did, RECORDS.r1.desiredContent) as SignedRecordDTO

    await context.ipfs.dag.put(signedRecord, FAKE_CID_2)

    const sPayload = dagCBOR.util.deserialize(signedRecord.linkedBlock)
    await context.ipfs.dag.put(sPayload, signedRecord.jws.link)

    // apply signed
    state = await tileDoctypeHandler.applyRecord(signedRecord.jws, FAKE_CID_2, context, state)

    await context.ipfs.dag.put(RECORDS.proof, FAKE_CID_4)
    // apply anchor
    state = await tileDoctypeHandler.applyRecord(RECORDS.r2.record as AnchorRecord, FAKE_CID_3, context, state)
    expect(state).toMatchSnapshot()
  })

  it('Does not apply anchor record on unsupported chain', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    const genesisRecord = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(genesisRecord, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(genesisRecord.linkedBlock)
    await context.ipfs.dag.put(payload, genesisRecord.jws.link)

    // apply genesis
    const state = await tileDoctypeHandler.applyRecord(genesisRecord.jws, FAKE_CID_1, context)

    // Create anchor proof with a different chainId than what's in the genesis record
    await context.ipfs.dag.put({ blockNumber: 123456, chainId: 'thewrongchain'}, FAKE_CID_4)
    // apply anchor
    await expect(tileDoctypeHandler.applyRecord(RECORDS.r2.record as AnchorRecord, FAKE_CID_3, context, state))
        .rejects.toThrow("Anchor proof chainId 'thewrongchain' is not supported. Supported chains are: 'fakechain:123'")
  })
})
