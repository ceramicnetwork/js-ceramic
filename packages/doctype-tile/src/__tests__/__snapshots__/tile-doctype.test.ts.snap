// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileDoctypeHandler applies anchor record correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
    "chainId": "fakechain:123",
  },
  "anchorStatus": 3,
  "content": Object {
    "much": "data",
    "very": "content",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 1,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          119,
          169,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 2,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
    ],
  },
  "signature": 2,
}
`;

exports[`TileDoctypeHandler applies genesis record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "much": "data",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
    ],
  },
  "signature": 2,
}
`;

exports[`TileDoctypeHandler applies signed record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "much": "data",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 1,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
    ],
  },
  "next": Object {
    "content": Object {
      "much": "data",
      "very": "content",
    },
    "metadata": Object {
      "controllers": Array [
        "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
      ],
    },
  },
  "signature": 2,
}
`;

exports[`TileDoctypeHandler squashes record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "test": "data",
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          119,
          169,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 1,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
    ],
  },
  "next": Object {
    "content": Object {
      "other": Object {
        "obj2": "fefe",
      },
    },
    "metadata": Object {
      "controllers": Array [
        "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
      ],
    },
  },
  "signature": 2,
}
`;
